src/v/redpanda/admin_server.cc

```diff
@@ -2646,67 +2698,119 @@ admin_server::unclean_abort_partition_reconfig_handler(
 }
 
 ss::future<ss::json::json_return_type>
-admin_server::set_partition_replicas_handler(
+admin_server::force_set_partition_replicas_handler(
   std::unique_ptr<ss::http::request> req) {
-    auto ntp = parse_ntp_from_request(req->param);
+    if (unlikely(!_controller->get_feature_table().local().is_active(
+          features::feature::force_partition_reconfiguration))) {
+        throw ss::httpd::bad_request_exception(
+          "Feature not active yet, upgrade in progress?");
+    }
 
+    auto ntp = parse_ntp_from_request(req->param);
     if (ntp == model::controller_ntp) {
         throw ss::httpd::bad_request_exception(
           fmt::format("Can't reconfigure a controller"));
     }
 
-    // make sure to call reset() before each use
-    static thread_local json::validator set_replicas_validator(
-      make_set_replicas_validator());
-
     auto doc = parse_json_body(*req);
-    apply_validator(set_replicas_validator, doc);
+    auto replicas = co_await validate_set_replicas(
+      doc, _controller->get_topics_frontend().local());
 
-    std::vector<model::broker_shard> replicas;
-    if (!doc.IsArray()) {
-        throw ss::httpd::bad_request_exception("Expected array");
-    }
-    for (auto& r : doc.GetArray()) {
-        const auto& node_id_json = r["node_id"];
-        const auto& core_json = r["core"];
-        if (!node_id_json.IsInt() || !core_json.IsInt()) {
-            throw ss::httpd::bad_request_exception(
-              "`node_id` and `core` must be integers");
+    const auto& topics = _controller->get_topics_state().local();
+    const auto& in_progress = topics.updates_in_progress();
+    const auto in_progress_it = in_progress.find(ntp);
+
+    if (in_progress_it != in_progress.end()) {
+        const auto& update = (*in_progress_it).second;
+        const auto state = update.get_state();
+        if (state != cluster::reconfiguration_state::in_progress) {
```

Comment:
would be great to support cancellations as well (and not that hard I think?), because I guess it will be common for users to first try to cancel normally and only then reach for force reconfiguration